chronos/asyncsync.nim

```diff
@@ -145,22 +146,19 @@ proc newAsyncEvent*(): AsyncEvent =
   result.waiters = newSeq[Future[void]]()
   result.flag = false
 
-proc removeWaiter(event: AsyncEvent, waiter: Future[void]) {.inline.} =
-  ## Removes ``waiter`` from list of waiters in ``lock``.
-  event.waiters.delete(event.waiters.find(waiter))
-
-proc wait*(event: AsyncEvent) {.async.} =
+proc wait*(event: AsyncEvent): Future[void] =
   ## Block until the internal flag of ``event`` is `true`.
   ## If the internal flag is `true` on entry, return immediately. Otherwise,
   ## block until another task calls `fire()` to set the flag to `true`,
   ## then return.
+  var w = newFuture[void]("AsyncEvent.wait")
+
   if not(event.flag):
-    var w = newFuture[void]("AsyncEvent.wait")
     event.waiters.add(w)
-    try:
-      await w
-    finally:
-      event.removeWaiter(w)
+  else:
```

Comment:
This procedure could be cancelled, please return `try/finally` block.

Reply:
it's no longer `async` so cancellations don't apply (this way we avoid a roundtrip to the event loop and all the extra cancellation future code)